stress: failed test in cockroach/server/server.test: TestStatusSummaries

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/04c996e9326c5093be30f4111ce8127e30fc312a

Stress build found a failed test:

```
=== RUN   TestStatusSummaries
E160826 04:58:42.702646 util/envutil/env.go:154  error parsing COCKROACH_LINEARIZABLE: strconv.ParseBool: parsing "abcd": invalid syntax
E160826 04:58:42.702663 util/envutil/env.go:154  error parsing COCKROACH_CONSISTENCY_CHECK_PANIC_ON_FAILURE: strconv.ParseBool: parsing "abcd": invalid syntax
E160826 04:58:42.702673 util/envutil/env.go:210  error parsing COCKROACH_MAX_OFFSET: time: invalid duration abcd
E160826 04:58:42.702684 util/envutil/env.go:210  error parsing COCKROACH_METRICS_SAMPLE_INTERVAL: time: invalid duration abcd
E160826 04:58:42.702692 util/envutil/env.go:210  error parsing COCKROACH_SCAN_INTERVAL: time: invalid duration abcd
E160826 04:58:42.702701 util/envutil/env.go:210  error parsing COCKROACH_SCAN_MAX_IDLE_TIME: time: invalid duration abcd
E160826 04:58:42.702710 util/envutil/env.go:210  error parsing COCKROACH_TIME_UNTIL_STORE_DEAD: time: invalid duration abcd
E160826 04:58:42.702718 util/envutil/env.go:210  error parsing COCKROACH_CONSISTENCY_CHECK_INTERVAL: time: invalid duration abcd
E160826 04:58:42.702728 util/envutil/env.go:154  error parsing COCKROACH_RESERVATIONS_ENABLED: strconv.ParseBool: parsing "abcd": invalid syntax
I160826 04:58:42.702751 storage/engine/rocksdb.go:354  opening in memory rocksdb instance
W160826 04:58:42.705521 gossip/gossip.go:1070  not connected to cluster; use --join to specify a connected node
W160826 04:58:42.706095 server/status/runtime.go:109  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I160826 04:58:42.706688 server/node.go:407  store store=0:0 not bootstrapped
I160826 04:58:42.709012 storage/replica_trigger.go:308  store=1:1 range=1 [/Min-/Max): new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 408940h58m51.708453315s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=2016-08-26 04:58:42.708969304 +0000 UTC]
I160826 04:58:42.709878 server/node.go:339  **** cluster {bf80f537-2181-43cb-9307-cae4d4473923} has been created
I160826 04:58:42.709905 server/node.go:340  **** add additional nodes by specifying --join=127.0.0.1:39366
I160826 04:58:42.710679 storage/store.go:982  store=1:1: failed initial metrics computation: store=1:1: system config not yet available
I160826 04:58:42.710708 server/node.go:420  initialized store store=1:1: {Capacity:536870912 Available:536870912 RangeCount:0}
I160826 04:58:42.710734 server/node.go:312  node ID 1 initialized
I160826 04:58:42.710800 storage/stores.go:302  read 0 node addresses from persistent storage
I160826 04:58:42.710865 server/node.go:549  connecting to gossip network to verify cluster ID...
I160826 04:58:42.710888 server/node.go:570  node connected via gossip and verified as part of cluster {"bf80f537-2181-43cb-9307-cae4d4473923"}
I160826 04:58:42.710923 server/node.go:358  node=1: started with [[]=] engine(s) and attributes []
I160826 04:58:42.710946 server/server.go:459  starting https server at 127.0.0.1:43470
I160826 04:58:42.710959 server/server.go:460  starting grpc/postgres server at 127.0.0.1:39366
I160826 04:58:42.711056 sql/event_log.go:109  Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:39366} Attrs:} ClusterID:{UUID:bf80f537-2181-43cb-9307-cae4d4473923} StartedAt:1472187522710903177}
I160826 04:58:42.711414 storage/split_queue.go:99  splitting store=1:1 range=1 [/Min-/Max) at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
I160826 04:58:42.713267 storage/replica_command.go:2213  store=1:1 range=1 [/Min-/Max): initiating a split of this range at key /Table/11
I160826 04:58:42.727285 storage/replica_command.go:2213  store=1:1 range=2 [/Table/11-/Max): initiating a split of this range at key /Table/12
I160826 04:58:42.734621 storage/replica_command.go:2213  store=1:1 range=3 [/Table/12-/Max): initiating a split of this range at key /Table/13
W160826 04:58:42.734748 storage/intent_resolver.go:344  could not GC completed transaction: end key /Local/Range/"\x93"/"rdsc\x00" must be greater than start /Local/Range/"\x93"/RangeDescriptor
I160826 04:58:42.738420 storage/replica_command.go:2213  store=1:1 range=4 [/Table/13-/Max): initiating a split of this range at key /Table/14
W160826 04:58:42.738736 storage/intent_resolver.go:344  could not GC completed transaction: end key /Local/Range/"\x94"/"rdsc\x00" must be greater than start /Local/Range/"\x94"/RangeDescriptor
W160826 04:58:42.742262 storage/intent_resolver.go:344  could not GC completed transaction: end key /Local/Range/"\x95"/"rdsc\x00" must be greater than start /Local/Range/"\x95"/RangeDescriptor
I160826 04:58:42.765451 sql/event_log.go:109  Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:39366} Attrs:} ClusterID:{UUID:bf80f537-2181-43cb-9307-cae4d4473923} StartedAt:1472187522710903177}
I160826 04:58:42.767549 sql/event_log.go:109  Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:39366} Attrs:} ClusterID:{UUID:bf80f537-2181-43cb-9307-cae4d4473923} StartedAt:1472187522710903177}
I160826 04:58:42.784213 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:39366->127.0.0.1:46484: use of closed network connection
--- FAIL: TestStatusSummaries (0.08s)
	node_test.go:401: 2: livebytes is not greater than or equal to expected value.
		Expected 22790.000000, Actual 22359.000000
	node_test.go:401: 2: valbytes is not greater than or equal to expected value.
		Expected 21194.000000, Actual 20735.000000
```

Run Details:

```
0 runs so far, 0 failures, over 5s
14 runs so far, 0 failures, over 10s
16 runs so far, 0 failures, over 15s
32 runs so far, 0 failures, over 20s
32 runs so far, 0 failures, over 25s
48 runs so far, 0 failures, over 30s
48 runs so far, 0 failures, over 35s
64 runs so far, 0 failures, over 40s
64 runs so far, 0 failures, over 45s
80 runs so far, 0 failures, over 50s
80 runs so far, 0 failures, over 55s
96 runs so far, 0 failures, over 1m0s
96 runs so far, 0 failures, over 1m5s
112 runs so far, 0 failures, over 1m10s
112 runs so far, 0 failures, over 1m15s
128 runs so far, 0 failures, over 1m20s
128 runs so far, 0 failures, over 1m25s
141 runs completed, 1 failures, over 1m27s
FAIL
```
Please assign, take a look and update the issue accordingly.

Finally got a repro on this with the necessary logging. For posterity, I will record what happened, as it illustrates some of the unexpected background events that occur on the server. Here is the log output, outfitted with some ad-hoc logging I used to investigate:

```
W160909 14:38:19.425947 99 storage/replica.go:2286  [n1,s1,r3{/Table/12-/Max}] !!! command *roachpb.EndTransactionRequest, valbytes diff -1395
I160909 14:38:19.426219 99 storage/store.go:1476  !!! updating MVCC Stats from split, valbytes diff 926
I160909 14:38:19.427625 117 storage/replica_command.go:2195  [n1,s1,r4{/Table/13-/Max}] initiating a split of this range at key /Table/14
W160909 14:38:19.430490 104 storage/replica.go:2286  [n1,s1,r1{/Min-/Table/11}] !!! command *roachpb.ResolveIntentRequest, valbytes diff -156
W160909 14:38:19.432047 107 storage/replica.go:2286  [n1,s1,r1{/Min-/Table/11}] !!! command *roachpb.NoopRequest, valbytes diff +203
W160909 14:38:19.442979 109 storage/replica.go:2286  [n1,s1,r4{/Table/13-/Max}] !!! command *roachpb.ConditionalPutRequest, valbytes diff +784
W160909 14:38:19.444761 110 storage/replica.go:2286  [n1,s1,r4{/Table/13-/Max}] !!! command *roachpb.EndTransactionRequest, valbytes diff -469
I160909 14:38:19.445084 110 storage/store.go:1476  !!! updating MVCC Stats from split, valbytes diff 0
W160909 14:38:19.447190 112 storage/replica.go:2286  [n1,s1,r1{/Min-/Table/11}] !!! command *roachpb.ResolveIntentRequest, valbytes diff -156
begin write statuses
W160909 14:38:19.527180 98 storage/replica.go:2286  [n1,n1,s1,r3{/Table/12-/Table/13}] !!! command *roachpb.BeginTransactionRequest, valbytes diff +702
I160909 14:38:19.527829 125 sql/event_log.go:109  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:56102} Attrs:} ClusterID:{UUID:58ae6a41-36a9-4256-beae-3182c6d1fbc2} StartedAt:1473446299297339812}
end write statuses
grabbing node statuses
W160909 14:38:19.531056 106 storage/replica.go:2286  [n1,n1,s1,r3{/Table/12-/Table/13}] !!! command *roachpb.ConditionalPutRequest, valbytes diff +702
W160909 14:38:19.532387 107 storage/replica.go:2286  [n1,n1,s1,r3{/Table/12-/Table/13}] !!! command *roachpb.EndTransactionRequest, valbytes diff -1178
W160909 14:38:19.547135 108 storage/replica.go:2286  [n1,s1,r1{/Min-/Table/11}] !!! command *roachpb.PutRequest, valbytes diff +9
W160909 14:38:19.547664 109 storage/replica.go:2286  [n1,s1,r1{/Min-/Table/11}] !!! command *roachpb.PutRequest, valbytes diff +9
begin write statuses
end write statuses
grabbing node statuses
I160909 14:38:19.573991 35 util/stop/stopper.go:342  stop has been called, stopping or quiescing all running tasks
I160909 14:38:19.591811 132 /Users/matt/go/src/google.golang.org/grpc/clientconn.go:472  Failed to dial 127.0.0.1:56102: connection error: desc = "transport: context canceled"; please retry.
--- FAIL: TestStatusSummaries (0.97s)
	node_test.go:407: 2: livebytes is not greater than or equal to expected value.
		Expected 24037.000000, Actual 23606.000000
	node_test.go:407: 2: valbytes is not greater than or equal to expected value.
		Expected 22441.000000, Actual 21982.000000
FAIL
```

This test inspects the value of status summaries (which are derived from metrics) several times in succession, and expects them to either be the same or, in some cases, greater than the previous value.  However, in this rare failure, the `valbytes` and `livebytes` fields decreased between two successive summaries.

My initial thought was that this was related to the phenomenon of metrics and splits: when stats are recalculated during a split, if the split range contained metrics data, the apparent number of bytes might decrease.  However, this was not the case here.

What *is* happening was made clear by logging every time valbytes was modified, along with the command that modified it.  I also logged when status summaries were generated.  As you can see above, the failure occurs when a transaction unexpectedly overlaps with the summary generation - because the transaction record still exists when the summary is generated, the `valbytes` count is unexpectedly higher when the summary is generated; the bytes immediately decreases when the transaction is finished, but by then the test's measurements have already been thrown off.

The transaction itself occurs on table 12, which is the Event log.  There is also a log message for event "node_join".  Sure enough, the "node_join" message is logged in a retryable background loop (to make a best effort to record this event for all nodes which join the cluster).  In this case, this random background transaction has overlapped our status generation, causing the anomaly.

The fix will be to either turn off the event log for this test, or to relax the expectations of the successive status calls.

--- a/base/test_server_args.go
+++ b/base/test_server_args.go
@@ -62,6 +62,9 @@ type TestServerArgs struct {
 	// Stopper can be used to stop the server. If not set, a stopper will be
 	// constructed and it can be gotten through TestServerInterface.Stopper().
 	Stopper *stop.Stopper
+
+	// If set, the recording of events to the event log tables is disabled.
+	DisableEventLog bool
 }
 
 // TestClusterArgs contains the parameters one can set when creating a test

--- a/server/context.go
+++ b/server/context.go
@@ -60,6 +60,7 @@ const (
 	defaultTimeUntilStoreDead       = 5 * time.Minute
 	defaultStorePath                = "cockroach-data"
 	defaultReservationsEnabled      = true
+	defaultEventLogEnabled          = true
 
 	minimumNetworkFileDescriptors     = 256
 	recommendedNetworkFileDescriptors = 5000
@@ -165,6 +166,12 @@ type Context struct {
 	// Ctx is the base context.Context for the server. If nil,
 	// context.Background() will be used.
 	Ctx context.Context
+
+	// EventLogEnabled is a switch which enables recording into cockroach's SQL
+	// event log tables. These tables record transactional events about changes
+	// to cluster metadata, such as DDL statements and range rebalancing
+	// actions.
+	EventLogEnabled bool
 }
 
 // GetTotalMemory returns either the total system memory or if possible the
@@ -337,6 +344,7 @@ func MakeContext() Context {
 		MetricsSampleInterval:    defaultMetricsSampleInterval,
 		TimeUntilStoreDead:       defaultTimeUntilStoreDead,
 		ReservationsEnabled:      defaultReservationsEnabled,
+		EventLogEnabled:          defaultEventLogEnabled,
 		Stores: StoreSpecList{
 			Specs: []StoreSpec{{Path: defaultStorePath}},
 		},

--- a/server/node_test.go
+++ b/server/node_test.go
@@ -441,12 +441,13 @@ func compareNodeStatus(t *testing.T, ts *TestServer, expectedNodeStatus *status.
 // both the Node and stores within the node.
 func TestStatusSummaries(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	t.Skip("#8844")
 
 	// ========================================
 	// Start test server and wait for full initialization.
 	// ========================================
-	srv, _, kvDB := serverutils.StartServer(t, base.TestServerArgs{})
+	srv, _, kvDB := serverutils.StartServer(t, base.TestServerArgs{
+		DisableEventLog: true,
+	})
 	defer srv.Stopper().Stop()
 	ts := srv.(*TestServer)
 

--- a/server/server.go
+++ b/server/server.go
@@ -260,7 +260,7 @@ func NewServer(srvCtx Context, stopper *stop.Stopper) (*Server, error) {
 		SQLExecutor: sql.InternalExecutor{
 			LeaseManager: s.leaseMgr,
 		},
-		LogRangeEvents: true,
+		LogRangeEvents: s.ctx.EventLogEnabled,
 		AllocatorOptions: storage.AllocatorOptions{
 			AllowRebalance: true,
 		},

--- a/server/testserver.go
+++ b/server/testserver.go
@@ -118,6 +118,9 @@ func makeTestContextFromParams(params base.TestServerArgs) Context {
 	if params.SSLCertKey != "" {
 		ctx.SSLCertKey = params.SSLCertKey
 	}
+	if params.DisableEventLog {
+		ctx.EventLogEnabled = false
+	}
 	ctx.JoinList = []string{params.JoinAddr}
 	return ctx
 }

